### DISCUSSION TITLE
Maintain own docker image for crossbuild

### DISCUSSION THREAD

**Discussion by dietmarw** [OpenModelica/OpenModelica#10459] [Ideas]
While trying to get #10411 to work one of the issues is that the docker image multiarch/crossbuild is out of date and one needs to additional translation flag due to the old cmake version. Now There already exists a simple PR https://github.com/multiarch/crossbuild/pull/61 to upgrade the docker image to a newer Debian version. Unfortunately, it looks like the development on multiarch/crossbuild has stagnated and that simple PR simply sits there without getting merged. 

So since OMEdit uses that docker image as part of one of the core features of the FMU export, would it not be a good idea to provide your own docker image that simply takes that update into account? 

That way the OpenModelica project has also lightly more control over the components it is dependent upon.

**Comment by dietmarw** [comment]
Sorry one oversight, it would not be enough to update to Debian 11 since OMC needs cmake >=3.21 and Debian 11 only comes with 3.18.4 **BUT** the beauty of having your own docker image is that you could base it on Debian 12

**Comment by dietmarw** [comment]
From #10462 it became also apparent that having its own custom docker image opens up for placing cross-compilation dependencies like for SUNDIALS and KLU  inside the docker image rather than OpenModelica installation is a benefit too. The OpenModelica installation is smaller and those who need the cross-compilation option can still get that out of the box.

**Comment by AnHeuermann** [comment]
I like the idea of having our own Docker images for cross-compilation. It could solve the issues with the dependency to CMake >= 3.21 and we could include some of the 3rd party dependencies into the images.
But I'm not sure if would call cross-compilation export of FMUs a core feature. But I don't have any way to tell how many users actually use this. I know that there are some questions about the "normal" FMU export every now and then.

From my point of view there are a three main tasks to set this up:

1. Create up-to-date Docker images with compiler tool-chains.
   - Create a fork of https://github.com/multiarch/crossbuild and update it.
   - Add some CI to test that it can compile a minimal C program for each target system. Maybe using CMake for that.
2. Compile 3rd party tools using the new image (could be the CI to test if the image works)
   - KLU
   - BLAS / LAPACK
   - SUNDIALS (CVODE)
   - ...
3. Testing FMU cross-compilation export 
    - Cross-compile FMUs from one host system to other target system(s).
4. Keeping it alive over the next decades
    - Someone has to be responsible to update everything and keep the CI fed and happy.

If someone can provide a solution for step 1 that would be great. Even without step 2 and 3 this could improve the FMU export experience immensely and we can simplify our CMake export (no backwards compatibility needed).
And of course keeping it alive is a never ending task which would use up some of our resources.

Hosting the image might be an issue as well. Docker changed the free team subscriptions, so we would need to pay, apply for the open source program or host them ourselves. If we want to be a reliable and industry capable tool we should have verified and checked images hosted somewhere official.
@casella @sjoelund do we want to try to apply for the [Docker open source program](https://www.docker.com/community/open-source/application/)? Worst case they say no, which is the status we have right now.

**Reply by dietmarw** [reply]
Just a short update, the free hosting is no longer an issue since Docker has learned of its errors: https://www.docker.com/blog/no-longer-sunsetting-the-free-team-plan/

**Reply by sjoelund** [reply]
An alternative to the huge crossbuild image would be to have a larger set of images available

**Reply by casella** [reply]
> But I'm not sure if would call cross-compilation export of FMUs a core feature.

There is for sure a huge market for that.

> But I don't have any way to tell how many users actually use this. 

Probably not that many, until we give production-quality support. Classic chicken-and-egg problem.

**Comment by AnHeuermann** [comment]
Another option could be to use https://github.com/dockcross/dockcross
